docs/abi.rst

```diff
@@ -0,0 +1,919 @@
+
+.. code-block:: python
+
+    from typing import Literal as L
+    from pyteal import *
+
+    @Subroutine(TealType.none)
+    def ensure_all_values_greater_than_5(array: abi.StaticArray[abi.Uint64, L[10]]) -> Expr:
+        """This subroutine asserts that every value in the input array is greater than 5."""
+        i = ScratchVar(TealType.uint64)
+        return For(
+            i.store(Int(0)), i.load() < array.length(), i.store(i.load() + Int(1))
+        ).Do(
+            array[i.load()].use(lambda value: Assert(value.get() > Int(5)))
+        )
+
+.. _Reference Types:
+
+Reference Types
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Many applications require the caller to provide "foreign array" values when calling the app. These are the blockchain entities (such as accounts, assets, or other applications) that the application will interact with when executing this call. In the ABI, we have **Reference Types** to describe these requirements.
+
+Definitions
+""""""""""""""""""""""""""""""""""""""""""
+
+PyTeal supports the following Reference Types:
+
+====================== ====================== ================ =======================================================================================================================================================
+PyTeal Type            ARC-4 Type             Dynamic / Static Description
+====================== ====================== ================ =======================================================================================================================================================
+:any:`abi.Account`     :code:`account`        Static           Represents an account that the current transaction can access, stored in the :any:`Txn.accounts <TxnObject.accounts>` array
+:any:`abi.Asset`       :code:`asset`          Static           Represents an asset that the current transaction can access, stored in the :any:`Txn.assets <TxnObject.assets>` array
+:any:`abi.Application` :code:`application`    Static           Represents an application that the current transaction can access, stored in the :any:`Txn.applications <TxnObject.applications>` array
+====================== ====================== ================ =======================================================================================================================================================
+
+These types all inherit from the abstract class :any:`abi.ReferenceType`.
+
+Limitations
+""""""""""""""""""""""""""""""""""""""""""
+
+Because References Types have a special meaning, they should not be directly created, and they cannot be assigned a value by a program.
+
+Additionally, Reference Types are only valid in the arguments of a method. They may not appear in a method's return value.
+
+Note that the AVM has limitations on the maximum number of foreign references an application call transaction may contain. At the time of writing, these limits are:
```

Comment:
@jasonpaulos Optional:  While it's convenient to list reference limits, I'm concerned it'll grow stale.  How open are you to one of the following alternatives?

* Link to a (or several) Github search(es) helping the user find the current value in `config/consensus.go`.
* Generally pointing the user to `config/consensus.go` to deduce values.

If you'd rather keep as is, then I don't feel strongly enough to pursue further and the thread can be resolved.

Reply:
Good point. I'm opting to link to https://developer.algorand.org/docs/get-details/parameter_tables/#smart-contract-constraints since it'll be more user-friendly than the actual code

Reply:
@jasonpaulos Re-opening for a final sanity check.  If the PR is in the expected form, then disregard.

From your prior note, I expected the hard-coded values to be removed.  It's unclear to me if you left them deliberately or incidentally.  

If incidental, also consider the same treatment in transaction type limitations (https://github.com/algorand/pyteal/pull/400/files#diff-d5f6d88ad06456a3ba42bcda3171b8d55309a3d87ad250a7e78379a322527d63R510) where max transaction group count is referenced.

Reply:
I've opted to leave the hard-coded values because I think they give a feel for the order of magnitude of the limits, plus they help illustrate the relationship between the per-field limits and the total reference limit. I hope that saying these values are only current at the time of writing and linking to a more authoritative source will be enough to future-proof this section.

Reply:
@jasonpaulos Got it - thanks for the clarification.  Resolving - I'm open to proceeding as is.